rust/connlib/tunnel/src/lib.rs

```diff
@@ -226,12 +262,29 @@ where
             gateway_public_keys,
         }
     }
+
+    pub async fn next_event(&self) -> Event<TIceState::Id> {
+        std::future::poll_fn(|cx| self.poll_next_event(cx)).await
+    }
+
+    pub fn poll_next_event(&self, cx: &mut Context<'_>) -> Poll<Event<TIceState::Id>> {
+        let (conn_id, candidate) = ready!(self.ice_state.lock().poll_next_ice_candidate(cx));
```

Comment:
For events generated from the handler loops(such as `connection_intent` from `iface_handler` or now I'm working on a dns fallback), do you expect to follow a same pattern? we poll here for different `StreamMap`s?

Reply:
See my latest changes. I think the way we want to go is to add more stuff to the different role-states and emit the events accordingly. We should only modify the state where types are known though. That will help us separate functions such that they are not seemingly generic whilst they are only used in one place.